integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala

```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
-             logicalRelation.relation.asInstanceOf[CarbonDatasourceHadoopRelation].carbonRelation.
-               metaData.hasAggregateDataMapSchema =>
-          val (carbonTable, tableName) = getCarbonTableAndTableName(logicalRelation)
-          isValidPlan = extractQueryColumnsFromAggExpression(groupingExp,
-            aggregateExp,
-            carbonTable,
-            tableName,
-            list)
-          if(isValidPlan) {
-            list ++
-            extractQueryColumnForOrderBy(Some(projectList), sortOrders, carbonTable, tableName)
-          }
-          carbonTable
-        // case for handling aggregation, order by and filter
-        case Project(projectList,
-          Sort(sortOrders,
-            _,
-            Aggregate(groupingExp,
-              aggregateExp,
-              Filter(filterExp, CarbonSubqueryAlias(_, logicalRelation: LogicalRelation)))))
-          if logicalRelation.relation.isInstanceOf[CarbonDatasourceHadoopRelation] &&
-             logicalRelation.relation.asInstanceOf[CarbonDatasourceHadoopRelation].carbonRelation.
-               metaData.hasAggregateDataMapSchema =>
-          val (carbonTable, tableName) = getCarbonTableAndTableName(logicalRelation)
-          isValidPlan = extractQueryColumnsFromAggExpression(groupingExp,
-            aggregateExp,
-            carbonTable,
-            tableName,
-            list)
-          if(isValidPlan) {
-            isValidPlan = !CarbonReflectionUtils.hasPredicateSubquery(filterExp)
+  /**
+   * Below method will be used to update the expression like group by expression
+   * @param expressions
+   * sequence of expression like group by
+   * @return updated expressions
+   */
+  def updateExpression(expressions : Seq[Expression]) : Seq[Expression] = {
+    val newExp = expressions map { expression =>
+      expression transform {
+        case attr: AttributeReference =>
+          val childExp = updatedExpression.find { p => p._1.sameRef(attr) }
+          if (childExp.isDefined) {
+            val newExpression = AttributeReference(
+              childExp.get._2.name,
+              childExp.get._2.dataType,
+              childExp.get._2.nullable,
+              childExp.get._2.metadata)(childExp.get._2.exprId, attr.qualifier, attr.isGenerated)
+            newExpression
+          } else {
+            attr
           }
-          if (isValidPlan) {
-            list ++
-            extractQueryColumnForOrderBy(Some(projectList), sortOrders, carbonTable, tableName)
-            isValidPlan = extractQueryColumnFromFilterExp(filterExp, list, carbonTable, tableName)
+      }
+    }
+    newExp
+  }
+
+  /**
+   * Below method will be used to updated the named expression like aggregate expression
+   * @param namedExpression
+   * any named expression like aggregate expression
+   * @return updated named expression
+   */
+  def updateNamedExpression(namedExpression: Seq[NamedExpression]) : Seq[NamedExpression] = {
```

Comment:
You can directly use `updateExpression` no need of this method. `NamedExpression` is derived from `Expression`. So after updation you can type cast.